utils/metrics/metrics.go

```diff
@@ -0,0 +1,187 @@
+package metrics
+
+import (
+	"github.com/buchgr/bazel-remote/config"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"strings"
+)
+
+// TODO Add test cases for this file
+
+type Method int
+type Status int
+type Kind int
+type Protocol int
+
+const (
+	METHOD_GET Method = iota
+	METHOD_HEAD
+	METHOD_PUT
+	METHOD_OTHER
+)
+
+const (
+	OK Status = iota
+	NOT_FOUND
+	OTHER_STATUS
+)
+
+const (
+	AC Kind = iota
+	CAS
+)
+
+const (
+	HTTP Protocol = iota
+	GRPC
+)
+
+func (e Method) String() string {
+	// Actually HTTP names, but can be conceptually mapped also to GRPC protocol.
+	if e == METHOD_GET {
+		return "GET"
+	}
+	if e == METHOD_HEAD {
+		return "HEAD"
+	}
+	if e == METHOD_PUT {
+		return "PUT"
+	}
+	return "other"
+}
+
+func (e Status) String() string {
+	// Names that works for both HTTP and GRPC, instead of HTTP or GRPC specific codes.
+	if e == OK {
+		return "OK"
+	}
+	if e == NOT_FOUND {
+		return "NotFound"
+	}
+	return "other"
+}
+
+func (e Kind) String() string {
+	if e == AC {
+		return "AC"
+	}
+	if e == CAS {
+		return "CAS"
+	}
+	return "other"
+}
+
+type Metrics interface {
+	// TODO Document interface
+	IncomingRequestCompleted(kind Kind, method Method, status Status, headers map[string][]string, protocol Protocol)
+}
+
+type metrics struct {
+	categoryValues               map[string]map[string]struct{}
```

Comment:
I wonder if we should consider simplifying this, and just use a list of allowed labels? eg `ubuntu1804_ci_trunk` ?

Reply:
Isn’t there a risk that would complicate it rather than simplifying?
I.e. introducing special cases like: ”ubuntu18” when only ”ubuntu1804” is allowed, handling underscore in values, etc?

However, I’m certainly not a fan of the golang syntax `map[string]struct{}` for a set of strings...

Reply:
I don't quite understand this map, how would it help if a client says ”ubuntu18” instead of ”ubuntu1804” ?

I figure that a list is simpler to implement, and probably easier to understand, at the cost of needing to enumerate all possible labels. But since we don't want too many labels, that number shouldn't be large anyway.

Reply:
I think I missinterpreted your example of "ubuntu1804_ci_trunk" as a string with underscore as separator. That's why I mentioned handling of underscore etc.

categoryValues is a map, with label names/header names (e.g. "os" and "user") as keys. The keys are mapped to values by the map, and each such value is a set of strings, containing allowed values for that label/header (e.g. {"ubuntu18-04", "ubuntu16-04"})

Perhaps categoryValues isn't a good name for such a map?

See also my long comment from today where I’m trying to elaborate on use cases.

Does it make sense?